### PR TITLE
Making redirectUri of OAuth2Settings/Provider Optional

### PR DESCRIPTION
**Pull request by Saheb**
# Pull Request Checklist

* [ ] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://github.com/mohiva/play-silhouette/blob/master/CONTRIBUTING.md)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/2.5.x/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [ ] Have you suggest documentation edits?
* [x] Have you added tests for any changed functionality?

## Fixes

Fixes #xxxx

To make it possible to make authorization requests without "redirect_uri" parameter, which even makes the model compliant with the OAuth2 specification.

What does this PR do?

It makes redirectUri parameter optional while making authorization requests to SocialAuth providers.

Why did you take this approach?

## References

Are there any relevant issues / PRs / mailing lists discussions?

Saheb Motiani @Saheb Nov 13 13:07
http://www.silhouette.rocks/docs/config-oauth2-based-providers
In OAuth2Settings, wouldn't it be good to make the redirectUrl an option isn't of having it as mandatory field?
Right now, this makes it impossible to make an authorization request without this parameter

Christian Kaps @akkie Nov 13 14:21
@SunPj Can you provide a repository which shows the issue?

Christian Kaps @akkie Nov 13 14:26
@Saheb What's your use case. Normally the redirectUrl is always needed for authentication. Every OAuth2 provider asks for the redirectUrlwhen creating the client ID and secret.

Saheb Motiani @Saheb Nov 13 14:30
Agreed, but it can use the one specified on Application's page too
if not specified in the authorization request

Saheb Motiani @Saheb Nov 13 14:36
For instance - https://developer.github.com/v3/oauth/#redirect-urls

Christian Kaps @akkie Nov 13 14:49
The spec defines it also as optional. Would you create a pull request?

Saheb Motiani @Saheb Nov 13 14:51
Sure, I can do that

Christian Kaps @akkie Nov 13 14:53

### CODE REVIEW

**Comment by Saheb** [comment]
@akkie As discussed on the mailing list, I have created simple PR with the required change. I have tested it manually with providing None for redirectUri in context. Let me know if you find a need to replicate test for all providers with a missing redirectUri and some tips on how to achieve that, without duplicating the existing test cases code.

File: silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala [hunk 1]
```diff
@@ -292,14 +299,4 @@ trait OAuth2StateProvider {
...
-  accessTokenURL: String,
-  redirectURL: String,
-  apiURL: Option[String] = None,
-  clientID: String,
-  clientSecret: String,
-  scope: Option[String] = None,
-  authorizationParams: Map[String, String] = Map.empty,
-  accessTokenParams: Map[String, String] = Map.empty,
-  customProperties: Map[String, String] = Map.empty)
+case class OAuth2Settings(authorizationURL: Option[String] = None, accessTokenURL: String, redirectURL: Option[String], apiURL: Option[String] = None, clientID: String, clientSecret: String, scope: Option[String] = None, authorizationParams: Map[String, String] = Map.empty, accessTokenParams: Map[String, String] = Map.empty, customProperties: Map[String, String] = Map.empty)
```
Comment (akkie) [inline silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/custom/FacebookProviderSpec.scala [hunk 2]
```diff
@@ -144,13 +144,7 @@ class FacebookProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://graph.facebook.com/oauth/authorize"),
-      accessTokenURL = "https://graph.facebook.com/oauth/access_token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = Some("email")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://graph.facebook.com/oauth/authorize"), accessTokenURL = "https://graph.facebook.com/oauth/access_token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = Some("email")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/custom/FacebookProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/Auth0ProviderSpec.scala [hunk 3]
```diff
@@ -148,14 +148,7 @@ class Auth0ProviderSpec extends OAuth2ProviderSpec {
...
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://gerritforge.eu.auth0.com/authorize"),
-      accessTokenURL = "https://gerritforge.eu.auth0.com/oauth/token",
-      apiURL = Some("https://gerritforge.eu.auth0.com/userinfo"),
-      redirectURL = "https://www.mohiva.com",
-      clientID = "some.client.id",
-      clientSecret = "some.secret",
-      scope = Some("email")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://gerritforge.eu.auth0.com/authorize"), accessTokenURL = "https://gerritforge.eu.auth0.com/oauth/token", redirectURL = Some("https://www.mohiva.com"), apiURL = Some("https://gerritforge.eu.auth0.com/userinfo"), clientID = "some.client.id", clientSecret = "some.secret", scope = Some("email")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/Auth0ProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/ClefProviderSpec.scala [hunk 4]
```diff
@@ -148,11 +148,7 @@ class ClefProviderSpec extends OAuth2ProviderSpec {
     /**
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      accessTokenURL = "https://clef.io/api/v1/authorize",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret"))
+    lazy val oAuthSettings = spy(OAuth2Settings(accessTokenURL = "https://clef.io/api/v1/authorize", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret"))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/ClefProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/DropboxProviderSpec.scala [hunk 5]
```diff
@@ -159,13 +159,7 @@ class DropboxProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://www.dropbox.com/1/oauth2/authorize"),
-      accessTokenURL = "https://api.dropbox.com/1/oauth2/token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = None))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://www.dropbox.com/1/oauth2/authorize"), accessTokenURL = "https://api.dropbox.com/1/oauth2/token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = None))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/DropboxProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/FacebookProviderSpec.scala [hunk 6]
```diff
@@ -151,13 +151,7 @@ class FacebookProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://graph.facebook.com/oauth/authorize"),
-      accessTokenURL = "https://graph.facebook.com/oauth/access_token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = Some("email")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://graph.facebook.com/oauth/authorize"), accessTokenURL = "https://graph.facebook.com/oauth/access_token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = Some("email")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/FacebookProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/FoursquareProviderSpec.scala [hunk 7]
```diff
@@ -210,12 +210,7 @@ class FoursquareProviderSpec extends OAuth2ProviderSpec {
     /**
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://foursquare.com/oauth2/authenticate"),
-      accessTokenURL = "https://foursquare.com/oauth2/access_token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret"))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://foursquare.com/oauth2/authenticate"), accessTokenURL = "https://foursquare.com/oauth2/access_token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret"))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/FoursquareProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/GitHubProviderSpec.scala [hunk 8]
```diff
@@ -149,13 +149,7 @@ class GitHubProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://github.com/login/oauth/authorize"),
-      accessTokenURL = "https://github.com/login/oauth/access_token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = Some("repo,gist")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://github.com/login/oauth/authorize"), accessTokenURL = "https://github.com/login/oauth/access_token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = Some("repo,gist")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/GitHubProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/GitLabProviderSpec.scala [hunk 9]
```diff
@@ -148,13 +148,7 @@ class GitLabProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://gitlab.com/oauth/authorize"),
-      accessTokenURL = "https://gitlab.com/oauth/token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = Some("api")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://gitlab.com/oauth/authorize"), accessTokenURL = "https://gitlab.com/oauth/token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = Some("api")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/GitLabProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/GoogleProviderSpec.scala [hunk 10]
```diff
@@ -169,13 +169,7 @@ class GoogleProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://accounts.google.com/o/oauth2/auth"),
-      accessTokenURL = "https://accounts.google.com/o/oauth2/token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = Some("profile,email")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://accounts.google.com/o/oauth2/auth"), accessTokenURL = "https://accounts.google.com/o/oauth2/token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = Some("profile,email")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/GoogleProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/InstagramProviderSpec.scala [hunk 11]
```diff
@@ -148,13 +148,7 @@ class InstagramProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://api.instagram.com/oauth/authorize"),
-      accessTokenURL = "https://api.instagram.com/oauth/access_token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = Some("basic")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://api.instagram.com/oauth/authorize"), accessTokenURL = "https://api.instagram.com/oauth/access_token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = Some("basic")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/InstagramProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/LinkedInProviderSpec.scala [hunk 12]
```diff
@@ -153,13 +153,7 @@ class LinkedInProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("https://www.linkedin.com/uas/oauth2/authorization"),
-      accessTokenURL = "https://www.linkedin.com/uas/oauth2/accessToken",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = None))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("https://www.linkedin.com/uas/oauth2/authorization"), accessTokenURL = "https://www.linkedin.com/uas/oauth2/accessToken", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = None))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/LinkedInProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/VKProviderSpec.scala [hunk 13]
```diff
@@ -167,13 +167,7 @@ class VKProviderSpec extends OAuth2ProviderSpec {
...
      * The OAuth2 settings.
      */
-    lazy val oAuthSettings = spy(OAuth2Settings(
-      authorizationURL = Some("http://oauth.vk.com/authorize"),
-      accessTokenURL = "https://oauth.vk.com/access_token",
-      redirectURL = "https://www.mohiva.com",
-      clientID = "my.client.id",
-      clientSecret = "my.client.secret",
-      scope = Some("email")))
+    lazy val oAuthSettings = spy(OAuth2Settings(authorizationURL = Some("http://oauth.vk.com/authorize"), accessTokenURL = "https://oauth.vk.com/access_token", redirectURL = Some("https://www.mohiva.com"), clientID = "my.client.id", clientSecret = "my.client.secret", scope = Some("email")))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/oauth2/VKProviderSpec.scala]: Please could you revert the formatting?

File: silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala [hunk 14]
```diff
@@ -142,12 +145,16 @@ trait OAuth2Provider extends SocialProvider with OAuth2Constants with Logger {
    * @return The info containing the access token.
    */
   protected def getAccessToken(code: String)(implicit request: RequestHeader): Future[OAuth2Info] = {
-    httpLayer.url(settings.accessTokenURL).withHeaders(headers: _*).post(Map(
+    val params = Map(
```
Comment (akkie) [inline silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala]: Maybe we could use the same schema as above, with the:

```scala
val redirectParam = settings.redirectURL match {
  case Some(rUri) => List((RedirectURI, resolveCallbackURL(rUri)))
  case None       => Nil
}
```

File: silhouette/test/com/mohiva/play/silhouette/impl/providers/OAuth2ProviderSpec.scala [hunk 15]
```diff
@@ -172,8 +175,11 @@ abstract class OAuth2ProviderSpec extends SocialProviderSpec[OAuth2Info] {
         ClientID -> Seq(c.oAuthSettings.clientID),
         ClientSecret -> Seq(c.oAuthSettings.clientSecret),
         GrantType -> Seq(AuthorizationCode),
-        Code -> Seq("my.code"),
-        RedirectURI -> Seq(c.oAuthSettings.redirectURL)) ++ c.oAuthSettings.accessTokenParams.mapValues(Seq(_))
+        Code -> Seq("my.code")) ++ c.oAuthSettings.accessTokenParams.mapValues(Seq(_))
```
Comment (akkie) [inline silhouette/test/com/mohiva/play/silhouette/impl/providers/OAuth2ProviderSpec.scala]: Same as above.

File: silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala (hunk 14 above)
Comment (Saheb) [inline silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala]: I did run the formatting script as mentioned in the contributing guide, not sure why this wasn't corrected by it. Any idea, what I am doing it wrong?

Comment (akkie) [inline silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala]: The script does not format such lines. Do you have formatted the code? Or does your IDE has formatted the code automatically?

Comment (Saheb) [inline silhouette/app/com/mohiva/play/silhouette/impl/providers/OAuth2Provider.scala]: Maybe my IDE did something, let me check it out.

**Comment by akkie** [comment]
Now some lines are formatted correctly and some lines are on one line. We need also a test which checks if the parameter was omitted. Then it looks good to merge!

**Comment by akkie** [comment]
Great! Thanks :+1:

**Pull request merged**